### DISCUSSION TITLE
Entity naming for devices with multiple channels

### DISCUSSION THREAD

**Discussion by thecode** [home-assistant/architecture#957] [Entity Models] [answered]
## Background
[Entity naming](https://developers.home-assistant.io/docs/core/entity/#entity-naming) introduced in https://github.com/home-assistant/developers.home-assistant/pull/1387 refer to 3 use cases for generating entity friendly name:

The `friendly_name` state attribute is generated by combining then entity name with the device name as follows:
- The entity is not a member of a device: `friendly_name = entity.name`
- The entity is a member of a device and `entity.name` is not `None`: `friendly_name = f"{device.name} {entity.name}"`
- The entity is a member of a device and `entity.name` is `None`: `friendly_name = f"{device.name}"`

## Scope:
The scope of this discussion is to expand the 2nd case, for devices that contains more than once channel and mainly (but not only) for devices that allow naming the channels. This discussion the assume is that handling for devices with a single channel is the optimal solution.
- The entity is a member of a device and `entity.name` is not `None`: `friendly_name = f"{device.name} {entity.name}"`

### Channel
A channel is a device that has duplication of a set of capabilities, mostly used by switches, but may also be in input devices or sensors. Channel may contain a main feature and multiple sensors that accompany this feature, common switch relay device with energy support may have the following

Main device entities such as `reboot`, `firmware`, `rssi`
Multiple channels with each channel repeating the same entities:
- Switch
- Current
- Power
- Energy
- Frequency
- Power Factor
- Voltage

## The problem
When [Entity naming](https://developers.home-assistant.io/docs/core/entity/#entity-naming) introduced it did not describe how to handle entities for devices that contains channels. Some integration owners assumed that everything after the device name belongs to the entity name, some try to split the device into multiple smaller devices were every channel is a device (which does not comply with the requirement that a device in Home Assistant is a physical device).
I think that the case for devices which contain channels, mainly channels that allow setting names was not fully addressed which left developers assuming how this should be implemented and made it harder for providing translations later. 
- How and if integrations should link multiple sensors which belongs to the same channel, some integrations embed the channel name in the entity name, some add an indexed channel name (`channel 1`,`channel 2`, `channel 3`) and some integrations doesn't handle this case at all, keeping all the sensors with the same name, making it complicated for the user to match the entities of a certain channel.
- How to translate entities if there is a channel name or a fixed text in the entity name (this can be solved by placeholders).
- How to handle devices with multiple channels that has different roles, this is mostly common for centralized installations which devices from integrations that are listed in the next section are often used. A an example for a dual role device is a switch with multiple channels that each channel control a different unrelated circuit. An example is a 4 relays switch device that has the following channels:
```
1. Master bedroom
2. Outdoor lights
3. Kitchen sockets
4. Gate
```
Appending a device name and embedding the channel name in the entity name creates a non-user friendly entities, assuming the device named `Left 4CH Relay` (due to its position in the main circuit box), entities will be named:
```
1. Left 4CH Relay Master bedroom
2. Left 4CH Relay Outdoor lights
3. Left 4CH Relay Kitchen sockets
4. Left 4CH Relay Gate
``` 
However, with a single channel device there is no problem, device will get its name by its role so the for the `Master bedroom` the device will be named `Master bedroom` and sub-entities will have the device name with the entity name appended (`Master bedroom power`, `Master bedroom energy`).
- Devices with high number of channels and energy monitoring usually ends up with large number of entities which is hard to show in  the frontend.

# Integrations that supports multiple channels
One of the most common questions in discussions is how many integrations benefit from this change, here the number of integrations is not large, but the integrations are very common integrations from the top of the used integrations and are one of the main roles for integrations used in Home Assistant for home control.
- [ESPHome](https://www.home-assistant.io/integrations/esphome/),
- [Shelly](https://www.home-assistant.io/integrations/shelly/)
- [Tasmota](https://www.home-assistant.io/integrations/tasmota/)
- [TP-Link Kasa Smart](https://www.home-assistant.io/integrations/tplink/)
- [KNX](https://www.home-assistant.io/integrations/knx/)
- [Z-Wave](https://www.home-assistant.io/integrations/zwave_js)
- [Zigbee Home Automation](https://www.home-assistant.io/integrations/zha)
- [MQTT](https://www.home-assistant.io/integrations/mqtt)

I am only using the first 3 so my information may be wrong regarding the others.

## Preferred solution:
When a device contain multiple channels a new attribute will store the channel name (for example `attr_channel`), the entity name will be formatted with one of the following options:

- The entity is a member of a device, `attr_channel` is not `None` and `entity.name` is not `None`: `friendly_name = f"{channel} {entity.name}"`

## Solution alternatives
- Don't deal with channels at all, cons are described in the **Problem** section of this discussion
- At an indexed property instead of a string based channel for grouping entities and signaling the frontend that entities belong to the same channel.
- Add an option for a sub-device, a sub device will not create a new device in HA but will be used to split the device into multiple smaller devices.

## Why should we do it?
- Allowing grouping multiple entities together, the frontend may use it when creating cards on the dashboard, so that instead of creating a single card for all the entities, the frontend will create card for each channel, bellow is an example of a 4 channel device with minimal enabled entities:
![image](https://github.com/home-assistant/architecture/assets/1858925/597321e7-1021-465f-b32f-d1b4c3a1631e)

It is already a non user friendly large card, while if it is splitted into multiple cards a card for a channel may look like this:
![image](https://github.com/home-assistant/architecture/assets/1858925/96608359-6c1a-482e-9188-aab512c24722)

- We will have deterministic way for handling channels, avoid implementations like this:
![image](https://github.com/home-assistant/architecture/assets/1858925/a82032f6-4383-409f-baa7-5f80d5ea2e86)
which leaves you with a bunch of entities not related to each other.
- Dashboard will be cleaner.
- We can support channel selection when viewing the device card in the frontend (for example a drop down selection) so that the device card is not overloaded with entities and managing the device will be easier.
- With the introduction of update entities, visit device and device cards improvements, many users (such as myself) replaced 3rd party management tools with Home Assistant capabilities and rely on the frontend for managing updates and device maintenance. Improving usability for multi-channel devices will extend this by many factors.
- Channel will be part of the `entity_id` in a determinist way, can be also expressed clearly as an entity attribute (for usage in the developer tools).

**I am sure I have missed many factors and considerations, but since this already too long I think it would be better to start the discussion and update it later with suggestions.**

**Comment by cgarwood** [comment]
Sub devices honestly feels the cleanest to me, and support mostly already exists. You can have a device set `via_device` to another device that has a `via_device` - and the UI already handles it relatively well IMO.

You have your hub device (`TP-Link Power Strip`) as a device that contains the overall sensors, like temperature, update entities, etc, then each plug or relay (`TV`, `Game Console`, `Christmas Tree`) would be a device with `via_device` set to the main `TP-Link Power Strip` device.

We don't have to introduce another layer of complexity, don't need to introduce a new "channel" concept. If we're concerned about Frontend, I wouldn't expect it would be difficult to indent sub-devices in the list or do some kind of tree view.

So my thought would be to change the "_every_ device must be a standalone physical device" policy, rather than adding additional complexity. I would still support a policy that says any device without a `via_device` set should be a standalone physical device.

**Reply by joostlek** [reply]
I think something to keep in mind is that the docs currently state "Although not currently available, we could consider offering an option to users to merge devices.".

When all plugs are in one device, it doesn't make sense to merge the device. But if you have your christmas tree Twinkly lights attached to the power brick (in a world the lights itself can only change color and not support turning on and off), you could combine both devices and merge this functionality into one device.

This isn't a major reason to do stuff, but I like this idea when I play around with it in my head.

**Reply by emontnemery** [reply]
I agree the sub device route seems promising, in particular because it avoids further complicating the already very complicated way in which the `friendly_name` is composed.

Also, I very much agree the problem raised is a real one which is worth spending effort on solving.

**Comment by MartinHjelmare** [comment]
Is channel the same as endpoint in the following description?

> A device that offers multiple endpoints, where parts of the device sense or output in different areas, should be split into separate devices and refer back to parent device with the via_device attribute. This allows the separate endpoints to be assigned to different areas in the building.

https://developers.home-assistant.io/docs/device_registry_index#what-is-a-device

**Reply by thecode** [reply]
There where strong arguments on a discord discussion that a physical device should not be split into multiple devices, what would be the name of each separate device? will it get its name from what I have defined as a channel in the discussion? 

For example, a common device - Shelly 2.5, which has common entities, 2 named channels and sensors for each channel. 
Assuming the device name is `Lights` and channel 1 name is `Bedroom` and channel 2 is `Kitchen` so if I understand it correctly the integration should create 3 devices:
1. Main device named `Lights` which contain common entities (e.g. `Firmware`, `rssi`, `reboot`)
2. A device named `Bedroom` which contain switch entity and related sensors (A switch named `Bedroom`, a sensor named `Bedroom power`, refers to 1 using `via_device`
3. Similar with a device named `Kitchen`

If this is correct than the discussion is probably answered with maybe small improvements for the frontend for displaying the relationship between the devices better.

**Reply by MartinHjelmare** [reply]
If channel corresponds to endpoint in the description above, then yes, we have guidelines how that should be handled. If we're not happy with those guidelines we'll need a new architecture discussion of course.

**Comment by joostlek** [comment]
I think there are some more cases that can be considered:

Unifi can also set a name for every port. To use translations in this point in time, it should create a device for each port.

QNAP wanted to add entity translations to the integration but they have sensors from 3 different types of devices, A physical hard disk, a volume and network ports. In this case, what do we count as a device? (cc @disforw)

Synology already has a separate device for disks and volumes.

System bridge can provide energy usage per core of your PC. Would a processor be a valid device? I have 32 cores, although the entities are disabled by default, I would love the device info page to be more categorized in parts of my pc for example. But do we count them as a different device or not.

Eightsleep is one device in theory, but it has two sides of a bed with the entity name containing both username and the side of the bed.

Freebox currently uses the raid id in entity names.

**Comment by AleXSR700** [comment]
I think that the forced addition of device name to every entity name is too self-centered. At least for Tasmota devices.

Most users will define friendly_names that are in themselves unique. Adding the device name to the friendly artificially creates long names which won't properly fit anymore and will often cause repetition.

The user has been taken the choice to use friendly_names in Tasmota properly and is no forced to chose friendly names that prevent repetition with the device name.
The consequence being, that any other client or system now receives strange friendly names because no one else forces device name + friendly name.

Let's take e.g. a device with one channel, e.g. a power socket.
I would call the device Stove_1 and use the same name for the friendly_name. That way, every client will receive the same clear information.

A device with 2 channels, e.g. a double power socket (Shelly 2.5 controlled), would have device name e.g. AV with friendly_name 1 being TV and friendly_name 2 being Amplifier.
Again, both are clear and standalone.
The new rules however force name the entities to sensor.av_tv or sensor.av_amplifier which makes absolutely no sense.

Why forcefully change a working system and make integrations that are not like others forcefully identical? Tasmota is meant to give the user freedom. Since when is it HA's goal to take away freedom?

Discussion also here https://github.com/home-assistant/core/issues/100247#issue-1893245418

**Comment by anonymousaga** [comment]
I need this to be reverted or made optional! Now all my dual-channel tasmota device names are messed up and get messed up every time i need to reinstall the tasmota integration!

**Comment by AleXSR700** [comment]
It seems that no one cares.
This was a bad idea and poor implementation. It messes up all devices with multiple entities.

Why not give the user the choice? Why force something so ridiculos on the users???

**Comment by frenck** [comment, marked as answer]
We have discussed this proposal and are currently declining this specific proposal.

We do think that most of these issue, right now, can be solved by adding support for placeholders in the entity name translations. This is thus, the path we like to take and explore first for now.

A PR for this has been created by @mib1185 in https://github.com/home-assistant/core/pull/104453

../Frenck
